File: javascript/10_summationOfPrimes.js
// The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
//
// Find the sum of all the primes below two million.

var summationOfPrimes = n => {
    var flags = [],
        sum = 2,
        limit = Math.sqrt(n);

    for (var i = 0; i <= n / 2; i++) flags.push(false, true);

    for (var i = 3; i <= limit; i++) {
        if (flags[i]) for (var k = i * i; k <= n; k += i) flags[k] = false;
    }

    for (var i = 3; i <= n; i++) {
        if (flags[i]) sum += i;
    }

    return sum;
}

console.log(summationOfPrimes(2000000));

File: ruby/10_summation_of_primes.rb
# The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
#
# Find the sum of all the primes below two million.

def summation_of_primes(n)
    sum, flags = 2, []

    0.upto(n / 2) { flags.push(false, true) }

    i = 3
    until i * i > n
        (i * i).step(n, i) { |k| flags[k] = false } if flags[i] == true
        i += 1
    end

    3.upto(n) { |j| sum += j if flags[j] == true }

    sum
end

p(summation_of_primes(2000000))
